sympy/physics/wigner.py

```diff
@@ -754,6 +762,158 @@ def gaunt(l_1, l_2, l_3, m_1, m_2, m_3, prec=None):
+        \begin{aligned}
+        \Theta(x) = \begin{cases} 1 \,\text{for}\, x > 0 \\ 0 \,\text{for}\, x \leq 0 \end{cases}
+        \end{aligned}
+
+    Parameters
+    ==========
+
+    l_1, l_2, l_3, m_1, m_2, m_3 :
+        Integer.
+    prec - precision, default: ``None``.
+        Providing a precision can
+        drastically speed up the calculation.
+
+    Returns
+    =======
+
+    Rational number times the square root of a rational number
+    (if ``prec=None``), or real number if a precision is given.
+
+    Examples
+    ========
+
+    >>> from sympy.physics.wigner import real_gaunt
+    >>> real_gaunt(2,2,4,-1,-1,0)
+    -2/(7*sqrt(pi))
+    >>> real_gaunt(10,10,20,-9,-9,0).n(64)
+    -0.00002480019791932209313156167...
+
+    It is an error to use non-integer values for `l` and `m`::
+
+        real_gaunt(2.8,0.5,1.3,0,0,0)
+        Traceback (most recent call last):
+        ...
+        ValueError: l values must be integer
+        real_gaunt(2,2,4,0.7,1,-3.4)
+        Traceback (most recent call last):
+        ...
+        ValueError: m values must be integer
+
+    Notes
+    =====
+
+    The real Gaunt coefficient inherits from the standard Gaunt coefficient, 
+    the invariance under any permutation of the pairs `(l_i, m_i)` and the 
+    requirement that the sum of the `l_i` be even to yield a non-zero value. 
+    It also obeys the following symmetry rules:
+
+    - zero for `l_1`, `l_2`, `l_3` not fulfiling the condition 
+    `l_1 \in \{l_{\text{max}}, l_{\text{max}}-2, \ldots, l_{\text{min}}\}`,
+    where `l_{\text{max}} = l_2+l_3`, 
+    
+    ..math::
+    
+        \begin{aligned}
+        l_{\text{min}} = \begin{cases} \kappa(l_2, l_3, m_2, m_3) & \text{if}\, 
+        \kappa(l_2, l_3, m_2, m_3) + l_{\text{max}}\, \text{is even} \\ 
+        \kappa(l_2, l_3, m_2, m_3)+1 & \text{if}\, \kappa(l_2, l_3, m_2, m_3) +
+        l_{\text{max}}\, \text{is odd}\end{cases}
+        \end{aligned}
+        
+    and `\kappa(l_2, l_3, m_2, m_3) = \max{\big(|l_2-l_3|, \min{\big(|m_2+m_3|, 
+    |m_2-m_3|\big)}\big)}`
+      
+    - zero for an odd number of negative `m_i`
+
+    Algorithms
+    ==========
+
+    This function uses the algorithms of [Homeier96]_ and [Rasch03]_ to
+    calculate the value of the real Gaunt coefficient exactly. Note that
+    the formula used in [Rasch03]_ contains alternating sums over large
+    factorials and is therefore unsuitable for finite precision arithmetic
+    and only useful for a computer algebra system [Rasch03]_. However, this
+    function can in principle use any algorithm that computes the Gaunt
+    coefficient, so it is suitable for finite precision arithmetic in so far 
+    as the algorithm which computes the Gaunt coefficient is.
+
+    """
+    if int(l_1) != l_1 or int(l_2) != l_2 or int(l_3) != l_3:
+        raise ValueError("l values must be integer")
+    if int(m_1) != m_1 or int(m_2) != m_2 or int(m_3) != m_3:
+        raise ValueError("m values must be integer")
```

Comment:
The reason I recommend using `as_int` is that it will give you a pure int (in case the user passed Integer) and it will raise the error for you. But again, if you require them to be ints then you don't allow the function to give a symbolic result. With this
```python
def real_gaunt(l_1,l_2,l_3,m_1,m_2,m_3,prec=None):
    step_func = lambda i: Piecewise((1,(i > 0)),(0,True))
    s = lambda e:  Piecewise((-1, Eq(Mod(e, 2), 1)), (1, True))
    U = lambda i,j: KroneckerDelta(abs(i), abs(j))*(KroneckerDelta(j, 0)*
             KroneckerDelta(i, 0)+step_func(i)*(KroneckerDelta(j, i)+s(j)*KroneckerDelta(j, -i)) + I*step_func(-i)*(-KroneckerDelta(j, i)*s(j-i)+KroneckerDelta(j, -i)*s(-i)))/sqrt(2)
    ugnt = []
    for i in range(-l_1, l_1+1):
        U1 = U(m_1, i)
        for j in range(-l_2, l_2+1):
            U2 = U(m_2, j)
            U3 = U(m_3, -i-j)
            ugnt.append(re(U1*U2*U3)*gaunt(l_1, l_2, l_3, i, j, -i-j))
    ugnt = Add(*ugnt)
    if prec is not None and ugnt.is_number:
        ugnt = ugnt.n(prec)
    return ugnt
```
you can get this
![image](https://user-images.githubusercontent.com/90703/189460516-bcb4d121-fd02-4fe5-a0d0-147628ff3a3e.png)
and that last expression can accept a substitution as
![image](https://user-images.githubusercontent.com/90703/189460579-14c1791c-cc90-4f64-b520-62b68ec53a86.png)

Reply:
So you might do this:
```python
try:
    m_1,m_2,m_3,l_1,l_2,l_3 = [as_int(i) for i in (m_1,m_2,m_3,l_1,l_2,l_3)]
    # call real_gaunt2
except ValueError:
    v = sympify((m_1,m_2,m_3,l_1,l_2,l_3))
    if not all(vi.is_integer for vi in v):
        raise ValueError('all parameters should be ints or integer expressions')
    # call symbolic version
```

Reply:
Hi Christopher,

my apologies for not responding, I was wrapped up in something else and forgot to reply to you. I really appreciate your code suggestions as you're right, they do make the program more concise and reduce the runtime substantially. I will implement your suggestions momentarily. I think we should leave the 'ugnt' variable as a scalar initially because I am concerned about memory usage (for large l values).

Real spherical harmonics are used a lot in quantum chemistry calculations, and I used them all throughout my master's research, so I think having a real gaunt function would be quite useful to others as well. I can solicit some other opinions however. 

I am just running some tests now but I have tested this algorithm in the past very extensively for my research. I have included some information about the real gaunt coefficient in the comments above the code. For reference, the l values can only be natural numbers. 

I mean symbolic in the same sense as the regular Gaunt algorithm, which only accepts integer arguments. I would wager that the symbolic case is not of interest to anyone, so I say we leave it out until people request it. What do you think?

Reply:
> I think we should leave the 'ugnt' variable as a scalar initially because I am concerned about memory usage (for large l values).

if it will not have many terms -- more than 1?? -- then that makes sense.

> tests

We just need enough tests to ensure that if anyone modifies the algorithm that we get catch errors. My first pass through, I had a different sign. So maybe just generating all results over a small range of integers would work:
```python
eigen = {(0, 0, 0, 0, 0, 0): 1/(2*sqrt(pi)), (2, 2, 0, 0, 0, 0): 1/(2*sqrt(pi)), (2, 0, 2, 1, 0, 1): 1/(2*sqrt(pi)), (1, 0, 1, 0, 0, 0): 1/(2*sqrt(pi)), (0, 1, 1, 0, -1, -1): 1/(2*sqrt(pi)), (0, 2, 2, 0, 1, 1): 1/(2*sqrt(pi)), (0, 1, 1, 0, 0, 0): 1/(2*sqrt(pi)), (1, 1, 0, -1, -1, 0): 1/(2*sqrt(pi)), (2, 0, 2, 2, 0, 2): 1/(2*sqrt(pi)), (0, 2, 2, 0, -2, -2): 1/(2*sqrt(pi)), (0, 2, 2, 0, -1, -1): 1/(2*sqrt(pi)), (0, 2, 2, 0, 0, 0): 1/(2*sqrt(pi)), (2, 2, 0, 1, 1, 0): 1/(2*sqrt(pi)), (2, 0, 2, 0, 0, 0): 1/(2*sqrt(pi)), (1, 0, 1, -1, 0, -1): 1/(2*sqrt(pi)), (2, 2, 0, -1, -1, 0): 1/(2*sqrt(pi)), (2, 2, 0, -2, -2, 0): 1/(2*sqrt(pi)), (0, 2, 2, 0, 2, 2): 1/(2*sqrt(pi)), (2, 0, 2, -2, 0, -2): 1/(2*sqrt(pi)), (2, 0, 2, -1, 0, -1): 1/(2*sqrt(pi)), (0, 1, 1, 0, 1, 1): 1/(2*sqrt(pi)), (1, 1, 0, 1, 1, 0): 1/(2*sqrt(pi)), (1, 0, 1, 1, 0, 1): 1/(2*sqrt(pi)), (2, 2, 0, 2, 2, 0): 1/(2*sqrt(pi)), (1, 1, 0, 0, 0, 0): 1/(2*sqrt(pi)), (2, 2, 2, 2, 2, 0): -sqrt(5)/(7*sqrt(pi)), (2, 2, 2, -2, -2, 0): -sqrt(5)/(7*sqrt(pi)), (2, 2, 2, 0, -2, -2): -sqrt(5)/(7*sqrt(pi)), (2, 2, 2, 2, 0, 2): -sqrt(5)/(7*sqrt(pi)), (2, 2, 2, -2, 0, -2): -sqrt(5)/(7*sqrt(pi)), (2, 2, 2, 0, 2, 2): -sqrt(5)/(7*sqrt(pi)), (1, 1, 2, 1, -1, -2): -sqrt(15)/(10*sqrt(pi)), (1, 2, 1, 1, -2, -1): -sqrt(15)/(10*sqrt(pi)), (2, 1, 1, -2, -1, 1): -sqrt(15)/(10*sqrt(pi)), (1, 1, 2, -1, 1, -2): -sqrt(15)/(10*sqrt(pi)), (1, 2, 1, -1, -2, 1): -sqrt(15)/(10*sqrt(pi)), (2, 1, 1, 2, -1, -1): -sqrt(15)/(10*sqrt(pi)), (1, 2, 1, -1, 2, -1): -sqrt(15)/(10*sqrt(pi)), (1, 1, 2, -1, -1, 2): -sqrt(15)/(10*sqrt(pi)), (2, 1, 1, -2, 1, -1): -sqrt(15)/(10*sqrt(pi)), (2, 2, 2, -1, 1, -2): -sqrt(15)/(14*sqrt(pi)), (2, 2, 2, -2, 1, -1): -sqrt(15)/(14*sqrt(pi)), (2, 2, 2, -1, 2, -1): -sqrt(15)/(14*sqrt(pi)), (2, 2, 2, 2, -1, -1): -sqrt(15)/(14*sqrt(pi)), (2, 2, 2, -1, -1, 2): -sqrt(15)/(14*sqrt(pi)), (2, 2, 2, -2, -1, 1): -sqrt(15)/(14*sqrt(pi)), (2, 2, 2, 1, -2, -1): -sqrt(15)/(14*sqrt(pi)), (2, 2, 2, -1, -2, 1): -sqrt(15)/(14*sqrt(pi)), (2, 2, 2, 1, -1, -2): -sqrt(15)/(14*sqrt(pi)), (2, 1, 1, 0, 1, 1): -sqrt(5)/(10*sqrt(pi)), (2, 1, 1, 0, -1, -1): -sqrt(5)/(10*sqrt(pi)), (1, 1, 2, -1, -1, 0): -sqrt(5)/(10*sqrt(pi)), (1, 1, 2, 1, 1, 0): -sqrt(5)/(10*sqrt(pi)), (1, 2, 1, -1, 0, -1): -sqrt(5)/(10*sqrt(pi)), (1, 2, 1, 1, 0, 1): -sqrt(5)/(10*sqrt(pi)), (2, 1, 1, 1, 0, 1): sqrt(15)/(10*sqrt(pi)), (1, 1, 2, 1, 1, 2): sqrt(15)/(10*sqrt(pi)), (2, 1, 1, 1, 1, 0): sqrt(15)/(10*sqrt(pi)), (1, 1, 2, -1, 0, -1): sqrt(15)/(10*sqrt(pi)), (2, 1, 1, -1, -1, 0): sqrt(15)/(10*sqrt(pi)), (1, 1, 2, 0, -1, -1): sqrt(15)/(10*sqrt(pi)), (1, 1, 2, 0, 1, 1): sqrt(15)/(10*sqrt(pi)), (1, 2, 1, 0, 1, 1): sqrt(15)/(10*sqrt(pi)), (1, 1, 2, 1, 0, 1): sqrt(15)/(10*sqrt(pi)), (1, 2, 1, 1, 2, 1): sqrt(15)/(10*sqrt(pi)), (1, 2, 1, -1, -1, 0): sqrt(15)/(10*sqrt(pi)), (2, 1, 1, 2, 1, 1): sqrt(15)/(10*sqrt(pi)), (2, 1, 1, -1, 0, -1): sqrt(15)/(10*sqrt(pi)), (1, 2, 1, 0, -1, -1): sqrt(15)/(10*sqrt(pi)), (1, 2, 1, 1, 1, 0): sqrt(15)/(10*sqrt(pi)), (2, 2, 2, -1, -1, 0): sqrt(5)/(14*sqrt(pi)), (2, 2, 2, 0, 1, 1): sqrt(5)/(14*sqrt(pi)), (2, 2, 2, 0, -1, -1): sqrt(5)/(14*sqrt(pi)), (2, 2, 2, 1, 1, 0): sqrt(5)/(14*sqrt(pi)), (2, 2, 2, -1, 0, -1): sqrt(5)/(14*sqrt(pi)), (2, 2, 2, 1, 0, 1): sqrt(5)/(14*sqrt(pi)), (2, 1, 1, 0, 0, 0): sqrt(5)/(5*sqrt(pi)), (1, 2, 1, 0, 0, 0): sqrt(5)/(5*sqrt(pi)), (1, 1, 2, 0, 0, 0): sqrt(5)/(5*sqrt(pi)), (2, 2, 2, 0, 0, 0): sqrt(5)/(7*sqrt(pi)), (2, 2, 2, 1, 1, 2): sqrt(15)/(14*sqrt(pi)), (2, 2, 2, 1, 2, 1): sqrt(15)/(14*sqrt(pi)), (2, 2, 2, 2, 1, 1): sqrt(15)/(14*sqrt(pi))}
for k in eigen:
    assert real_gaunt(*k) == eigen[k]
r = [-2]*6 + [-1]*6 + [0]*6 + [1]*6 + [2]*6
while True:
    shuffle(r)
    if r[:6] not in eigen:
        assert real_gaunt(*r[:6]) == 0  # 15625 possibilities
        break
```
Note: I do not know if those values are correct.

Reply:
> if it will not have many terms -- more than 1?? -- then that makes sense.

Typical calculations do not use that high an l value, but I have seen calculations where the l values reached up to 100.

> We just need enough tests to ensure that if anyone modifies the algorithm that we get catch errors. My first pass through, I had a different sign. So maybe just generating all results over a small range of integers would work:

I have tested the algorithm for the range `0 < l_1, l_2, l_3 < 10` and have generated a txt file containing all non-zero values. I can upload it somewhere in the repository so that it can be referred to in the event of modifications. 

[Non-Zero Real Gaunt Coefficients for l1, l2, l3 Between 0 and 10.txt](https://github.com/sympy/sympy/files/9541961/Non-Zero.Real.Gaunt.Coefficients.for.l1.l2.l3.Between.0.and.10.txt)